File: apps/driver-android/src/main/kotlin/lk/mageride/driver/wallet/LankaQrCode.kt
package lk.mageride.driver.wallet

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.WriterException
import com.google.zxing.common.BitMatrix
import com.google.zxing.qrcode.QRCodeWriter
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import lk.mageride.driver.ui.theme.MageRideTheme

/**
 * AL-15's **fallback**: the LankaQR payload rendered as a scannable code.
 *
 * The primary path is the *"Pay"* deep link into the driver's own bank app, and this is what
 * SCR-DA-022 shows when that link could not be opened ([PaymentHandoff] answers `false`) or when
 * the server sent a payload and no link at all. `PaymentMethods.lankaQrAction` in `:shared` is
 * where that choice is made; this composable only draws what it decided.
 *
 * **Always black on white, in both themes.** A QR code is read by a camera's contrast, not by a
 * design system — inverting it for dark mode produces a code that many scanners refuse, so the
 * quiet zone and the modules are fixed and the card around them is what the theme colours. This is
 * the one place in the module that names a colour outside `ui/theme`, and that is why.
 */
@Composable
internal fun LankaQrCode(payload: String, modifier: Modifier = Modifier) {
    val matrix = remember(payload) { lankaQrMatrix(payload) } ?: return

    Canvas(
        modifier = modifier
            .fillMaxWidth()
            .aspectRatio(1f)
            .clip(RoundedCornerShape(MageRideTheme.radius.md))
            .background(Color.White)
            .padding(MageRideTheme.spacing.sm),
    ) {
        val module = size.minDimension / matrix.width
        val moduleSize = Size(module, module)

        for (row in 0 until matrix.height) {
            for (column in 0 until matrix.width) {
                if (!matrix.get(column, row)) continue
                drawRect(
                    color = Color.Black,
                    topLeft = Offset(x = column * module, y = row * module),
                    size = moduleSize,
                )
            }
        }
    }
}

/**
 * The payload as a module grid, or `null` when it cannot be encoded.
 *
 * `null` rather than a throw: an EMVCo payload arrives from a gateway, and a malformed one must
 * leave the driver on a screen that still offers OnePay rather than take the app down. The margin
 * is zero because the composable draws its own quiet zone as padding on a white field — ZXing's
 * default four-module margin would be drawn *inside* the same square and halve the code.
 */
internal fun lankaQrMatrix(payload: String): BitMatrix? = try {
    QRCodeWriter().encode(
        payload,
        BarcodeFormat.QR_CODE,
        MODULE_GRID,
        MODULE_GRID,
        mapOf(
            // `M` is EMVCo's own recommendation for a payment code: `L` is fragile on a screen
            // held at an angle and `H` inflates a long payload into modules too fine to scan.
            EncodeHintType.ERROR_CORRECTION to ErrorCorrectionLevel.M,
            EncodeHintType.MARGIN to 0,
            // The payload is EMVCo TLV — ASCII by construction — and naming the charset stops
            // ZXing guessing a multi-byte one and adding an ECI header no scanner needs.
            EncodeHintType.CHARACTER_SET to "ISO-8859-1",
        ),
    )
} catch (_: WriterException) {
    null
} catch (_: IllegalArgumentException) {
    null
}

/**
 * The requested output size — **one**, which asks ZXing for the bare module grid.
 *
 * `QRCodeWriter` scales its symbol into whatever size is asked for and **centres it**, so a request
 * for 512 px produces a 512-wide matrix whose modules are 11 px each and whose first eight columns
 * are padding. Two things would then be wrong: the canvas below draws one `drawRect` per matrix
 * cell, which at 512² is a quarter of a million rectangles a frame, and the grid it is measuring
 * would not start at the finder pattern. Asking for 1 makes `outputWidth = max(1, symbolWidth)`
 * and the multiple 1, so a cell is a module and the canvas does the scaling — which is what a
 * vector surface should be doing anyway. Not a `dp` for exactly that reason.
 */
private const val MODULE_GRID = 1

File: apps/driver-ios/DriverApp/Wallet/LankaQrCode.swift
import CoreImage
import CoreImage.CIFilterBuiltins
import SwiftUI

/// AL-15's **fallback**: the LankaQR payload rendered as a scannable code.
///
/// The primary path is the *"Pay"* link into the driver's own bank app, and this is what SCR-DI-022
/// shows when that link could not be opened (``PaymentHandoff/openBankApp(_:)`` answers `false`) or
/// when the server sent a payload and no link at all. `PaymentMethods.lankaQrAction` in `:shared` is
/// where that choice is made; this view only draws what it decided.
///
/// **Δ iOS — the encoder is Core Image's and there is no dependency.** The Android twin added
/// `com.google.zxing:core` for this one job (and C074 later used its reader half for SCR-DA-027).
/// `CIQRCodeGenerator` is a first-party filter that has shipped since iOS 7, so the same fallback
/// costs this target nothing — which also means the AL-34 fence is easier to hold here than there:
/// **no code in this app can read a QR at all**, because nothing links a decoder.
///
/// **Always black on white, in both themes.** A QR code is read by a camera's contrast, not by a
/// design system — inverting it for dark mode produces a code many scanners refuse — so the modules
/// and the quiet zone are fixed and the card around them is what the theme colours. `CIQRCodeGenerator`
/// already emits black-on-white, so nothing here tints it either.
struct LankaQrCode: View {

    let payload: String

    var body: some View {
        Group {
            if let image = LankaQrCode.render(payload) {
                Image(uiImage: image)
                    // Nearest-neighbour, not the default smoothing: a QR symbol is a grid of hard
                    // squares and a bilinear upscale of a 30-module image is a grey blur no scanner
                    // reads. This is the whole reason the filter's tiny output is drawn rather than
                    // rasterised at display size.
                    .interpolation(.none)
                    .resizable()
                    .scaledToFit()
                    .padding(MageRideSpacing.sm)
                    .background(.white, in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous))
                    .accessibilityLabel(Text(key: "wallet_lankaqr_title"))
            } else {
                // A payload the encoder refused. The screen keeps its copy and its Close button rather
                // than an empty square: an EMVCo payload arrives from a gateway, and a malformed one
                // must leave the driver on a screen that still offers OnePay.
                Text(key: "wallet_lankaqr_unavailable")
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.error)
                    .multilineTextAlignment(.center)
                    .frame(maxWidth: .infinity)
            }
        }
        .frame(maxWidth: .infinity)
        .aspectRatio(1, contentMode: .fit)
    }

    /// The payload as a symbol, or `nil` when it cannot be encoded.
    ///
    /// `nil` rather than a fatal error, for the reason above. The filter answers `nil` for a payload
    /// too long to fit the format, and `CIContext.createCGImage` for an extent it cannot render.
    ///
    /// The correction level is **M**, EMVCo's own recommendation for a payment code: `L` is fragile on
    /// a screen held at an angle and `H` inflates a long payload into modules too fine to scan. The
    /// payload is EMVCo TLV — ASCII by construction — so it is encoded as ISO-8859-1, which stops the
    /// generator adding an ECI header no scanner needs.
    static func render(_ payload: String) -> UIImage? {
        guard let data = payload.data(using: .isoLatin1) else { return nil }

        let filter = CIFilter.qrCodeGenerator()
        filter.message = data
        filter.correctionLevel = "M"

        guard let output = filter.outputImage,
              let cgImage = context.createCGImage(output, from: output.extent)
        else { return nil }

        return UIImage(cgImage: cgImage)
    }

    /// One context for the app.
    ///
    /// A `CIContext` allocates a rendering pipeline, and building one per draw is the classic Core
    /// Image cost. The fallback dialog can be re-rendered on every Dynamic Type change.
    private static let context = CIContext()
}
